cpp/src/lists/contains.cu

```diff
@@ -44,384 +41,345 @@
-                                          rmm::cuda_stream_view stream)
-{
-  return column_device_view::create(search_keys, stream);
-}
+/**
+ * @brief A sentinel value used for marking that a given key has not been found in the search list.
+ */
+auto constexpr __device__ NOT_FOUND_SENTINEL = size_type{-1};
 
-auto get_search_keys_device_iterable_view(cudf::scalar const& search_key, rmm::cuda_stream_view)
+template <typename Type>
+auto constexpr is_supported_non_nested_type()
 {
-  return &search_key;
+  return cudf::is_fixed_width<Type>() || std::is_same_v<Type, cudf::string_view>;
 }
 
-template <typename ElementType, duplicate_find_option find_option>
-auto __device__ find_begin(list_device_view const& list)
-{
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return list.pair_rep_begin<ElementType>();
-  } else {
-    return thrust::make_reverse_iterator(list.pair_rep_end<ElementType>());
-  }
-}
-
-template <typename ElementType, duplicate_find_option find_option>
-auto __device__ find_end(list_device_view const& list)
+/**
+ * @brief Return a pair of index iterators {begin, end} to loop through elements within a list.
+ *
+ * Depending on the value of `find_first`, a pair of forward or reverse iterators will be returned,
+ * allowing to loop through elements in the list in first-to-last or last-to-first order.
+ *
+ * Note that the element indices always restart to `0` at the first position in each list.
+ *
+ * @tparam find_first A boolean value indicating whether we want to find the first or last
+ *         appearance of a given key in the list.
+ * @param size The number of elements in the list.
+ * @return A pair of {begin, end} iterators to iterate through the range `[0, size)` by forward or
+ *         reverse order.
+ */
+template <bool find_first>
+auto __device__ element_index_pair_iter(size_type const size)
 {
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return list.pair_rep_end<ElementType>();
+  if constexpr (find_first) {
+    return thrust::pair(thrust::make_counting_iterator<size_type>(0),
+                        thrust::make_counting_iterator<size_type>(size));
   } else {
-    return thrust::make_reverse_iterator(list.pair_rep_begin<ElementType>());
+    return thrust::pair(
+      thrust::make_reverse_iterator(thrust::make_counting_iterator<size_type>(size)),
+      thrust::make_reverse_iterator(thrust::make_counting_iterator<size_type>(0)));
   }
 }
 
-template <duplicate_find_option find_option, typename Iterator>
-size_type __device__ distance([[maybe_unused]] Iterator begin, Iterator end, Iterator find_iter)
-{
-  if (find_iter == end) {
-    return absent_index;  // Not found.
+/**
+ * @brief Functor for searching indices of the given key(s) within each list row in a list column.
+ */
+template <typename Type, typename Enable = void>
+struct search_index_fn {
+  template <typename... Args>
+  static void invoke(Args&&...)
```

Comment:
I hadn't considered this before. This is an interesting point. I was shown the `Args&&...` technique to prevent having to repeat the signature for the "everything else" part of a dispatch endpoint. Good catch.

Reply:
There will be two “free” template parameters for the iterators but you can constrain it to five arguments with specific types for the 1st, 4th, and 5th parameters. Just copy the template and types from the other one.

Reply:
I could make these `invoke` function standalone template functions instead. However, each of them depends on a separate `search_list` function so I decided to put them into different specializations of template struct.